PR: Add a call-to-action for Kite

metalogical opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes

This PR adds a call-to-action to educate users about Kite at a moment the user would benefit from Kite completions (only if no other completions are shown to the user). It is an improved version of #10192.
![Screen Shot 2019-10-07 at 6 06 18 PM](https://user-images.githubusercontent.com/1977419/66359766-36944e80-e92d-11e9-9eae-ef669a5f8df2.png)

We err on the side of minimizing false positives, so the dialog is not displayed in every case where Kite may provide completions, but only in cases where where we are highly confident Kite would produce sensible results.

The detection is implemented only for the following case, using regular expressions to identify the function that must be understood by Kite in order to provide completions:
```python
foo = full.path.to.function.that.kite.understands(...)
...
x = foo.bar.b‸
```
We then match this function against a pre-computed bloom filter containing common function paths that Kite understands.

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @metalogical 

<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @metalogical! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





















* In the file [`spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/__init__.py`](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/__init__.py):

> [Line 6:1](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/__init__.py#L6): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

* In the file [`spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py`](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py):

> [Line 42:80](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py#L42): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 107:80](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py#L107): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 163:25](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py#L163): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 165:80](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py#L165): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 79 characters)
> [Line 174:80](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py#L174): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 177:25](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py#L177): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 179:80](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py#L179): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 196:18](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py#L196): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator
> [Line 241:80](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py#L241): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 320:80](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py#L320): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 327:80](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/pybloom_pyqt/pybloom.py#L327): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)



* In the file [`spyder/utils/external/pyqt_bloomfilter/setup.py`](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/setup.py):

> [Line 7:80](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/setup.py#L7): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (284 > 79 characters)
> [Line 14:1](https://github.com/spyder-ide/spyder/blob/6c927b8fadc2617a8d986d86ddd8fda1cc24a5a9/spyder/utils/external/pyqt_bloomfilter/setup.py#L14): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

metalogical commented:
Now that the install UI is in, I'm updating this to trigger that. I'm also cleaning up some design elements (e.g. buttons instead of links)

metalogical commented on spyder/config/main.py:
```diff
@@ -92,6 +92,7 @@
               'completion/size': (300, 180),
               'report_error/remember_me': False,
               'report_error/remember_token': False,
+              'show_kite_cta': True,
```
We need a way of having "Dismiss Forever" state persist. Those options seem separate (e.g. Kite can be enabled by default, without Kite actually being installed, in which case the CTA should still show, until the user clicks "dismiss").

This is analogous to the `show_installation_dialog` option added for the purposes of dismissing the install dialog: https://github.com/spyder-ide/spyder/pull/10251/files#diff-33b5b1425c385f61c36db4a4f1350bf9R73

metalogical replied:
moving to the `kite` section of the config though.

metalogical replied:
oh nvm, I see what you're saying now 🤦‍♂ I added the setting there and here.

I think the right approach here is to address the merge in whichever PR ends up getting merged second.

metalogical commented on spyder/plugins/completion/kite/ui/cta.py:
```diff
@@ -0,0 +1,181 @@
+        # TODO Learn More & Install Kite should trigger the install flow UI
+        # and disable the CTA for the session
+        labels_layout.addWidget(self._create_link_label(
+            translate("Install Kite"), "https://kite.com/download/"))
```
Did you mean replace with the `_` way you suggested above (that's what I've done)? Or remove entirely? If the latter, I can knock that out whenever you say so.

metalogical commented:
@ccordoba12 I've addressed all your comments, excepting one question about translation, which I left inline.

Regarding your other concern: we autogenerated a list of around 13K function paths (fully qualified, and also a few common aliases) for which we understand the return types / semantics. We did some filtering, etc as well to make sure the list is not excessively large (no private functions, only functions exported in top-level packages, etc).

[paths.txt](https://github.com/spyder-ide/spyder/files/3692825/paths.txt)

The list still contains a decent amount of noise, which we could clean up a bit, but didn't expend the energy, since we're happy with the resulting size of the bloom filter.

ccordoba12 commented on spyder/plugins/completion/kite/widgets/calltoaction.py:
```diff
@@ -0,0 +1,151 @@
+        border_color = color_scheme['sideareas']
+        text_color, __, __ = color_scheme['normal']
+        button_color = color_scheme['currentline']
+        hover_color = color_scheme['currentcell']
```
I think using our color scheme for the CTA colors is the wrong thing to do. I mean `color_scheme['normal']` is fine, but the other colors vary wildly aming color schemes. For instance, using `color_scheme['background']` in light themes makes no sense because that color is `white`. And `color_scheme['currentline']` is a shade purple in other schemes.

@metalogical, I think you should use the same colors as the completion widget for background and foreground, and also preserve the same colors and hover behavior for buttons as any other widget.

metalogical replied:
done: I had to use `setFrameStyle` to force the frame style to match `QListWidget`, use `setAutoFillBackground` to prevent the frame from being transparent in the Light UI mode, and copied the window flags from the completion widget.

It still doesn't look 100% identical, but extremely close: the completions widget has rounded corners, and in light mode, has a blue highlight around the border. Not sure how that happens.

ccordoba12 commented:
> Regarding your other concern: we autogenerated a list of around 13K function paths (fully qualified, and also a few common aliases) for which we understand the return types / semantics

Nice, thanks for sharing those paths! That's really useful to understand for us to understand when the CTA is going to be displayed.

dalthviz commented on spyder/plugins/completion/kite/widgets/calltoaction.py:
```diff
@@ -0,0 +1,134 @@
+        super(KiteCallToAction, self).__init__(ancestor)
+        self.textedit = textedit
+
+        self.setFrameStyle(54)
```
Please change this to `self.setFrameStyle(QFrame.StyledPanel | QFrame.Plain)` (the equivalent using constants to the value `54`)

dalthviz commented on spyder/plugins/completion/kite/widgets/calltoaction.py:
```diff
@@ -0,0 +1,134 @@
+        self.setFrameStyle(54)
+        self.setAutoFillBackground(True)
+        self.setWindowFlags(Qt.SubWindow | Qt.FramelessWindowHint)
+        self.setFocusPolicy(Qt.NoFocus)
```
Please, below this line, add the following to have the correct behavior of the border when using the dark theme:
```python
    if is_dark_interface():
            self.setObjectName("kite-call-to-action")
            self.setStyleSheet(
                self.styleSheet() + ("#kite-call-to-action "
                                     "{ border: 1px solid; "
                                     "  border-color: #32414B; "
                                     "  border-radius: 4px;} "
                                     "#kite-call-to-action:hover "
                                     "{ border:1px solid #148CD2; }"))
```

You will need to add the import for `is_dark_interface` at the imports section of the file. The import is 

```python
from spyder.utils.icon_manager import is_dark_interface
```

metalogical replied:
lovely, this fixes it, thanks!

dalthviz commented on spyder/plugins/completion/kite/widgets/calltoaction.py:
```diff
@@ -0,0 +1,144 @@
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+from qtpy.QtCore import Qt, QTimer, QUrl
```
Also @naman I think there are some imports that are not needed anymore (like QTimer, QUrl, QDesktopServices, etc). Please remove them
